Code/src/main/java/nl/utwente/group10/haskell/type/ConstT.java

```diff
@@ -1,35 +1,73 @@
 package nl.utwente.group10.haskell.type;
 
+import com.google.common.base.Joiner;
+
 /**
- * A specific, non-variable Haskell type.
+ * Constant, concrete type. However, it may consist of variable types.
  */
 public class ConstT extends Type {
     /**
-     * The Haskell name of this type.
+     * The constructor for this type.
+     */
+    private final String constructor;
+
+    /**
+     * The types of the arguments for this type.
+     */
+    private final Type[] args;
+
+    /**
+     * @param constructor The constructor for this constant type.
+     * @param args The types of the arguments that this type accepts.
+     */
+    public ConstT(final String constructor, final Type ... args) {
+        this.constructor = constructor;
+        this.args = args;
+    }
+
+    /**
+     * @return The constructor for this type.
      */
-    private final String name;
+    public final String getConstructor() {
+        return this.constructor;
+    }
 
     /**
-     * @param name The Haskell name of this type.
+     * @return The types of the arguments for this type.
      */
-    public ConstT(final String name) {
-        this.name = name;
+    public final Type[] getArgs() {
+        return this.args;
     }
 
     @Override
-    public final boolean compatibleWith(final Type other) {
-        return this.equals(other);
+    public final Type prune() {
+
+        for (int i = 0; i < this.args.length; i++) {
+            this.args[i] = this.args[i].prune();
+        }
+
+        return this;
     }
 
     @Override
-    public final String toHaskellType() {
-        return this.name;
+    public String toHaskellType() {
+        StringBuilder out = new StringBuilder();
+        out.append(this.constructor);
+
+        for (int i = 0; i < this.args.length; i++) {
+            out.append(" ");
+            out.append(this.args[i].toHaskellType());
+        }
+
+        return out.toString();
     }
 
     @Override
-    public final String toString() {
-        return "ConstT{" +
-                "name='" + this.name + "'" +
-                "}";
+    public String toString() {
+        if (this.args.length > 0) {
+            return String.format("(%s %s)", this.constructor, Joiner.on(' ').join(this.args));
+        } else {
+            return this.constructor;
+        }
```

Comment:
In this specific case, you're implementing the same idea twice with entirely different approaches. If you don't want the methods to depend on each other, at least pick one and copy that :)